frontend/javascripts/oxalis/view/right-menu/tree_hierarchy_view_helpers.js

```diff
@@ -86,15 +88,28 @@ export function insertTreesAndTransform(
     const treeNode = makeTreeNodeFromGroup(group, {
       // Ensure that groups are always at the top when sorting by timestamp
       timestamp: 0,
-      expanded: expandedGroupIds[groupId] != null ? expandedGroupIds[groupId] : true,
+      expanded: expandedGroupIds[groupId] != null ? expandedGroupIds[groupId] : false,
       children: insertTreesAndTransform(group.children, groupToTreesMap, expandedGroupIds, sortBy),
     });
     // Groups are always sorted by name and appear before the trees, trees are sorted according to the sortBy prop
     const trees = _.orderBy(groupToTreesMap[groupId] || [], [sortBy], ["asc"]).map(
       makeTreeNodeFromTree,
     );
     treeNode.children = _.orderBy(treeNode.children, ["name"], ["asc"]).concat(trees);
-    treeNode.isChecked = _.every(treeNode.children, groupOrTree => groupOrTree.isChecked);
+    // Empty groups should show up as indeterminate and not checked
+    treeNode.isChecked =
+      treeNode.children.length > 0 &&
+      _.every(treeNode.children, groupOrTree => groupOrTree.isChecked);
+    treeNode.isIndeterminate = treeNode.isChecked
+      ? false
+      : treeNode.children.length === 0 ||
+        _.some(
+          treeNode.children,
+          groupOrTree =>
+            groupOrTree.isChecked ||
+            // Empty groups which are shown as indeterminate should not influence the indeterminate state of their parents
+            (groupOrTree.isIndeterminate && groupOrTree.children.length > 0),
+        );
```

Comment:
Whew, my brain is twisting :smile: Kudos for figuring this out!

I'm not sure if you are aware, but for the case that a group contains an empty group, this still affects the parent group and makes it indeterminate. See this screenshot:

![image](https://user-images.githubusercontent.com/2486553/97291052-ae3f5f00-1849-11eb-8b97-773681183dab.png)

Is this different from your described "Empty groups which are shown as indeterminate should not influence the indeterminate state of their parents"-case? I think the problem with the current logic is that the `treeNode.isChecked = ...` does not take into account when a child is indeterminate because it's an empty group. That means, as soon as there is an empty group as a child, the parent group cannot be checked (even though, we would like it to be unaffected by the indeterminate, empty child).

Maybe the following would work?
```suggestion
    treeNode.isChecked =
      treeNode.children.length > 0 &&
      _.every(treeNode.children, groupOrTree => groupOrTree.isChecked || (groupOrTree.isIndeterminate && groupOrTree.children.length > 0));
    treeNode.isIndeterminate = treeNode.isChecked
      ? false
      : treeNode.children.length === 0 ||
        _.some(
          treeNode.children,
          groupOrTree =>
            groupOrTree.isChecked ||
            // Empty groups which are shown as indeterminate should not influence the indeterminate state of their parents
            (groupOrTree.isIndeterminate && groupOrTree.children.length > 0),
        );
```

Reply:
Thanks for testing this! I think the correct fix would be

```javascript
_.every(
  treeNode.children,
  groupOrTree =>
    groupOrTree.isChecked ||
    // Empty groups which are shown as indeterminate should not influence the indeterminate state of their parents
    (groupOrTree.isIndeterminate && groupOrTree.children.length === 0),
);
```

(`=== 0` instead of `> 0`)

However, what about empty groups within groups. With the "fixed" code the parent group that only contains an empty group would always show up as "Checked", but following the logic it would also need to be "Indeterminate" :/

I feel like this is getting too complicated and we should just do it right, by saving the visibility information for each group in the store. As far as I can see this would resolve all these problems and empty groups would be assigned the correct visibility state instead of this "indeterminate" workaround. What do you think? Let's call if it's easier :)

Reply:
> I feel like this is getting too complicated and we should just do it right, by saving the visibility information for each group in the store. As far as I can see this would resolve all these problems and empty groups would be assigned the correct visibility state instead of this "indeterminate" workaround. What do you think?

Yes, you are probably right! However, that's probably out of scope for this PR, right? Then, feel free to remove the "show empty groups as indeterminate" commit again :)

Reply:
As discussed, I went with the method to hide the checkbox for all groups that do not contain any trees. I used the same indentation as if the checkbox was there which looks better than without the indentation in my opinion, but I'm open for suggestions :)

Reply:
Great! Works very well for me :100: I also tried how it looks without the indentation and I also think that it looks better with it. I think, the margin is still a tiny bit weird, but that's just looks and overall the best solution we have until now :) So, it's good to go :+1: 

By the way: I also checked how GIMP does this case:
![image](https://user-images.githubusercontent.com/2486553/97317983-adb7c000-186b-11eb-9956-85a1e00bed96.png)

The visibility icon is always at the very far left. Not sure whether this would work with our library. And for deep hierarchies, this might also be unpleasant to use due to the high margin between checkbox and tree/group name.

Regarding the data model: GIMP seems to store a visibility flag for all items, but then derives the actual visibility from that afterwards. So, a group can be invisible (no "eye") while the child is supposed to be visible ("eye"). However, the child's visibility is then overriden to be invisible (--> "eye" is crossed). This has the benefit, that toggling the parent does not "lose" the visibility-information of all children. Interesting approach, but I think, ours also works fine for now, too :)